### PR TITLE
[no ticket][risk=no] Puppeteer flaky tests fix

### PR DESCRIPTION
**Pull request by aweng98**
`workspace-share-modal.spec` has failed [here](https://app.circleci.com/pipelines/github/all-of-us/workbench/10494/workflows/7edb403d-69aa-4393-b504-006c715a576f/jobs/141162).
`dataset-snowman-menu-actions.spec.ts` has failed [here](https://app.circleci.com/pipelines/github/all-of-us/workbench/10498/workflows/866a4093-cf0b-4cf9-afff-a29489080fcf/jobs/141189).

Adding two whitespaces at the end of the user name has caused share-modal test to fail on master branch. The fix drops the whitespace and search for dropdown as puppeteer is typing char by char instead of type all char then search for dropdown.

### CODE REVIEW

File: e2e/app/modal/share-modal.ts [hunk 1]
```diff
@@ -21,31 +21,57 @@ export default class ShareModal extends Modal {
...
+    const typeAndAddUser = async () => {
+      await this.waitForSearchBox().type(userName, { delay: 50 });
+      await waitForDropDown();
+      try {
+        const addIcon = findCollaboratorAddIcon();
+        await addIcon.click();
+        await waitForClose();
+        return;
+      } catch (err) {
+        // Click failed.
```
Comment (yonghaoy) [inline e2e/app/modal/share-modal.ts:62]: I think it would be helpful to log exception here.

File: e2e/app/modal/share-modal.ts [hunk 2]
```diff
@@ -99,4 +117,19 @@ export default class ShareModal extends Modal {
     const levelText = level[0].toUpperCase() + level.substring(1).toLowerCase();
     return this.page.waitForXPath(`//*[starts-with(@id,"react-select") and text()="${levelText}"]`, { visible: true });
   }
+
+  async notFindUser(email: string): Promise<boolean> {
```
Comment (yonghaoy) [inline e2e/app/modal/share-modal.ts:121]: where&how this method is used?

File: e2e/tests/datasets/dataset-create.spec.ts [hunk 3]
```diff
@@ -9,8 +9,8 @@ import WorkspaceAboutPage from 'app/page/workspace-about-page';
 import { config } from 'resources/workbench-config';
 import { makeWorkspaceName } from 'utils/str-utils';
 
-// 30 minutes. Test involves notebook that could take a long time to create.
```
Comment (yonghaoy) [inline e2e/tests/datasets/dataset-create.spec.ts:12]: What is the reason to refuce this? Does notebook creation becomes faster?

File: e2e/tests/workspace/workspace-share-modal.spec.ts [hunk 4]
```diff
@@ -4,7 +4,7 @@ import WorkspaceAboutPage from 'app/page/workspace-about-page';
 import { config } from 'resources/workbench-config';
 import WorkspacesPage from 'app/page/workspaces-page';
 import WorkspaceDataPage from 'app/page/workspace-data-page';
-import { makeWorkspaceName } from 'utils/str-utils';
+//import { makeWorkspaceName } from 'utils/str-utils';
```
Comment (yonghaoy) [inline e2e/tests/workspace/workspace-share-modal.spec.ts:7]: delete it not needed?

File: e2e/tests/workspace/workspace-share-modal.spec.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ describe('Workspace Share Modal', () => {
   ];
 
   // Create new workspace with default CDR version
-  const workspace = makeWorkspaceName();
+  const workspace = 'aoutest-407581621629013'; //makeWorkspaceName();
```
Comment (yonghaoy) [inline e2e/tests/workspace/workspace-share-modal.spec.ts:16]: that is this?

**Review by yonghaoy — COMMENTED** [review]
Thanks! few comment/questions.

File: e2e/app/modal/share-modal.ts (hunk 2 above)
Comment (aweng98) [reply e2e/app/modal/share-modal.ts]: not used yet. I plan to use it in future in a new negative search test.

File: e2e/tests/datasets/dataset-create.spec.ts (hunk 3 above)
Comment (aweng98) [reply e2e/tests/datasets/dataset-create.spec.ts:12]: Comment was an error. None of tests in this file create or start a notebook.

File: e2e/tests/workspace/workspace-share-modal.spec.ts (hunk 5 above)
Comment (aweng98) [reply e2e/tests/workspace/workspace-share-modal.spec.ts]: Good catch. I forgot to undo the change. It's used for local testing.

File: e2e/app/modal/share-modal.ts (hunk 1 above)
Comment (aweng98) [reply e2e/app/modal/share-modal.ts]: I reworked code to search for add-icon as puppeteer type char by char.

**Comment by yonghaoy** [comment]
See slack message: https://pmi-engteam.slack.com/archives/C7JUTPQLE/p1621636402004500
I think the `two space` causes this flakiness, and we probably want to fix this from the root cause. Simply adding retries probably can reduce flakiness, but won't solve this problem. 
The value to having those test is to capture bugs in our code, not just make them add passing. In real world, if researchers have to retry few times while trying to share, then this is a perfect test that we capture this issue without retries. We may want to make it work by

1. No need to type the extra space as suffix
2. UI trim extra space suffix.
1 is a preferred option because that is a good user experience. Do you recall why we need the extra spaces? Probably just not enough time? Do you think it would work if we remove the two extra space, and just let it sleep for 100ms?

File: e2e/tests/datasets/export-to-notebook.spec.ts [hunk 6]
```diff
@@ -1,13 +1,13 @@
 import WorkspaceAnalysisPage from 'app/page/workspace-analysis-page';
 import WorkspaceDataPage from 'app/page/workspace-data-page';
 import NotebookPreviewPage from 'app/page/notebook-preview-page';
-import {makeRandomName, makeWorkspaceName} from 'utils/str-utils';
-import {findOrCreateWorkspace, signInWithAccessToken} from 'utils/test-utils';
+import { makeRandomName, makeWorkspaceName } from 'utils/str-utils';
```
Comment (aweng98) [inline e2e/tests/datasets/export-to-notebook.spec.ts:4]: changes done by `yarn lint --fix`.

File: e2e/app/modal/share-modal.ts [hunk 7]
```diff
@@ -21,33 +20,54 @@ export default class ShareModal extends Modal {
...
     };
     const waitForClose = async () => {
       await waitWhileLoading(this.page);
       await this.page.waitForXPath(dropDownXpath, { hidden: true });
     };
+    const findCollaboratorAddIcon = () => {
+      return ClrIconLink.findByName(
+        this.page,
+        { type: ElementType.Icon, iconShape: 'plus-circle', containsText: userName, ancestorLevel: 2 },
+        this
```
Comment (aweng98) [inline e2e/app/modal/share-modal.ts:42]: Ideally the search for icon should be done inside the dropdown object. but that change will causes more changes. TODO in a separate coming PR.

**Review by yonghaoy — APPROVED** [review]
Thanks!

**Pull request merged**